### PR TITLE
 Added custom metrics binding credential generation and management in APIServer

### PR DESCRIPTION
**Pull request by paltanmoy**
Custom Metrics feature related changes[For ServiceBroker and APIServer only].

### CODE REVIEW

**Comment by cdlliuy** [comment]
@paltanmoy ,   Is that possible to break this PR with 2930 insertions into smaller ones , i.e. to handle  credential generation and management only ?  Regardless custom metric,  credential generation is a "valid" topic in service broker and api server ..

File: src/integration/integration_broker_api_test.go [hunk 1]
```diff
@@ -211,6 +211,40 @@ var _ = Describe("Integration_Broker_Api", func() {
...
+				Consistently(fakeScheduler.ReceivedRequests).Should(HaveLen(1))
+
+				By("checking the API Server")
+				var expected map[string]interface{}
+				err = json.Unmarshal(schedulePolicyJson, &expected)
+				Expect(err).NotTo(HaveOccurred())
+
+				checkResponseContent(getPolicy, appId, http.StatusOK, expected, INTERNAL)
+			})
+		})
```
[the remark this answers is not in the document]
Comment (paltanmoy) [reply src/integration/integration_broker_api_test.go]: Added Unbind Test case

File: api/lib/routes/credentials.js [hunk 2]
```diff
@@ -0,0 +1,79 @@
+'use strict';
+module.exports = function(models, credentialCache, cacheTTL) {
+
+  var express = require('express');
+  var router = express.Router();
+  var logger = require('../log/logger');
+  var HttpStatus = require('http-status-codes');
+  var credHelper = require('./credentialHelper')(models, credentialCache, cacheTTL);
+
+  router.get('/:app_id/creds', function(req, resp) {
```
[the remark this answers is not in the document]
Comment (paltanmoy) [reply api/lib/routes/credentials.js]: Yes , all endpoints have been modified as suggested

File: api/lib/routes/credentialHelper.js [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+      logger.error('Failed to create custom metrics credentials', {
+        'app_id': req.params.app_id,
+        'error': error
+      });
+      error.statusCode = HttpStatus.INTERNAL_SERVER_ERROR;
+      callback(error);
+    });
+  }
+
+  credhelper.deleteCredentials = function(req, callback) {
```
[the remark this answers is not in the document]
Comment (paltanmoy) [reply api/lib/routes/credentialHelper.js]: Cache invalidation added.

File: api/lib/routes/credentialHelper.js [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+  }
+
+  credhelper.createOrUpdateCredentials = function(req, callback) {
+    var username = uuidv4();
+    var password = uuidv4();
+    models.credentials.upsert({
+      id: req.params.app_id,
+      username: generateHash(username),
+      password: generateHash(password)
+    }).then(function(updatedData) {
```
[the remark this answers is not in the document]
Comment (paltanmoy) [reply api/lib/routes/credentialHelper.js]: Nope, rather sequelize upsert returns a promise with a boolean indicating whether the data is created or not.  So renamed updatedData  to createdData

File: api/lib/routes/credentialHelper.js [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+      if (updatedData) {
+        logger.info('New credentials hasbeen generated successfully', {
```
[the remark this answers is not in the document]
Comment (paltanmoy) [reply api/lib/routes/credentialHelper.js]: Done

File: api/lib/routes/credentialHelper.js [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+      callback(insufficientQueryparamError);
+      return;
+    }
+    // Try to find credentials in cache
+    try{
+      creds = credentialCache.get(appId, true);
+      isValidCred = validateCredentialDetails(username, creds.username, password, creds.password);
+      logger.info('Credentials hasbeen found successfully in cache', {
+          'app_id': appId,
+          'isValid': isValidCred
```
Comment (boyang9527) [inline api/lib/routes/credentialHelper.js]: if it is not valid, we may need to check with database... and refresh cache.

**Comment by boyang9527** [comment]
LGTM

File: servicebroker/lib/util/apiServerUtil.js [hunk 7]
```diff
@@ -25,20 +25,78 @@ module.exports = function(settings) {
...
+  };
+
+  apiUtil.deleteCreds = function(appId, callback){
+    var options = credentialOptions(appId);
+    request.delete(options,function(error, response, body){
+      callback(error,response);
+    });
+  };
+
+  apiUtil.validateCreds = function(parameters, callback){
```
Comment (qibobo) [inline servicebroker/lib/util/apiServerUtil.js]: Is this function useful?

File: servicebroker/test/routes/binding_test.js [hunk 8]
```diff
@@ -112,14 +148,38 @@ describe('binding RESTful API', function() {
...
+            expect(res.body.credentials.custom_metrics.password).to.equal('newpassword');
+
             binding.count({ where: { bindingId: bindingId } }).then(function(countRes) {
               expect(countRes).to.equal(1);
               done();
             })
           });
       });
+
+      it("failed to create a new binding because of credential generation failure", function(done) {
```
Comment (qibobo) [inline servicebroker/test/routes/binding_test.js]: The binding creation only fails when enableCustomMetrics is true.

File: api/lib/routes/credentialHelper.js [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+    var isValidCred = isUsernameValid && isPasswordValid;
+    return isValidCred;
+  }
+
+  credhelper.createOrUpdateCredentials = function(req, callback) {
+    var username = uuidv4();
+    var password = uuidv4();
+    var appId = req.params.app_id;
+    models.credentials.upsert({
+      id: req.params.app_id,
```
[the remark this answers is not in the document]
Comment (paltanmoy) [reply api/lib/routes/credentialHelper.js]: which one? id => appId or app_id => appId ?

File: api/test/unit/config/settings_test.js [hunk 10]
```diff
@@ -87,6 +88,7 @@ describe('config setting Test Suite', function () {
     expect(settings.cfApi).to.equal("https://" + defaultConfig.cfApi.toLowerCase());
     expect(settings.infoFilePath).to.equal(defaultConfig.infoFilePath);
     expect(settings.skipSSLValidation).to.equal(false);
+    expect(settings.cacheTTL).to.equal(defaultConfig.cacheTTL);
```
[the remark this answers is not in the document]
Comment (paltanmoy) [reply api/test/unit/config/settings_test.js]: Test case added

File: api/lib/routes/credentialHelper.js [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+      logger.error('Failed to validate custom metrics credentials due to insufficient query parameters', {
+        'app_id': appId,
+        'error': insufficientQueryparamError
+      });
+      callback(insufficientQueryparamError);
+      return;
+    }
+    // Try to find credentials in cache
+    try{
+      creds = credentialCache.get(appId, true);
```
[the remark this answers is not in the document]
Comment (paltanmoy) [reply api/lib/routes/credentialHelper.js]: As getting creds directly from database is being already handled in catch block ..tried handling in the same way.

File: servicebroker/test/routes/binding_test.js (hunk 8 above)
Comment (paltanmoy) [reply servicebroker/test/routes/binding_test.js]: in test settings enableCustomMetrics is true only

File: api/lib/routes/credentialHelper.js [hunk 12]
```diff
@@ -0,0 +1,196 @@
...
+    var username = uuidv4();
+    var password = uuidv4();
+    var appId = req.params.app_id;
+    models.credentials.upsert({
+      id: req.params.app_id,
+      username: generateHash(username),
+      password: generateHash(password)
+    }).then(function(createdData) {
+      if (createdData) {
+        logger.info('New credentials hasbeen generated successfully', {
```
[the remark this answers is not in the document]
Comment (paltanmoy) [reply api/lib/routes/credentialHelper.js]: done

File: api/lib/routes/credentialHelper.js [hunk 13]
```diff
@@ -0,0 +1,196 @@
...
+          'app_id': appId
+        });
+        callback(null, {
+          'statusCode': HttpStatus.CREATED,
+          'username': username,
+          'password': password
+        });
+      }
+      else {
+        logger.info('Existing credentials hasbeen updated successfully', {
```
[the remark this answers is not in the document]
Comment (paltanmoy) [reply api/lib/routes/credentialHelper.js]: Done

File: servicebroker/lib/util/apiServerUtil.js (hunk 7 above)
Comment (paltanmoy) [reply servicebroker/lib/util/apiServerUtil.js]: Nope. Removed the function.

File: api/lib/routes/credentialHelper.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (qibobo) [reply api/lib/routes/credentialHelper.js]: var appId = req.params.app_id;
appId has been defined, so better to " id: appId" ?

**Comment by qibobo** [comment]
LGTM

**Pull request merged**